File: TrumpMobile/TrumpMobile/OrderFlow/OrderDetailView.swift
import Firebase
import FirebaseAuth
import FirebaseFirestore
import SwiftUI

struct OrderDetailView: View {
  let orderId: String
  @State private var orderDetail: OrderDetail?
  @State private var isLoading = true
  @State private var errorMessage: String?
  @Environment(\.dismiss) private var dismiss
  @EnvironmentObject private var navigationState: NavigationState

  private let dateFormatter: DateFormatter = {
    let formatter = DateFormatter()
    formatter.dateStyle = .medium
    formatter.timeStyle = .short
    return formatter
  }()

  var body: some View {
    NavigationView {
      ScrollView {
        if isLoading {
          ProgressView("Loading order details...")
            .frame(maxWidth: .infinity, maxHeight: .infinity)
            .padding(.top, 100)
        } else if let errorMessage = errorMessage {
          VStack {
            Image(systemName: "exclamationmark.triangle")
              .font(.largeTitle)
              .foregroundColor(.red)
            Text("Error")
              .font(.title2)
              .fontWeight(.semibold)
            Text(errorMessage)
              .multilineTextAlignment(.center)
              .padding()
          }
          .padding()
        } else if let order = orderDetail {
          VStack(spacing: 20) {
            // Order Header
            orderHeaderSection(order)

            // Personal Information
            personalInfoSection(order)

            // Address Information
            addressInfoSection(order)

            // Device Information
            deviceInfoSection(order)

            // Service Information
            serviceInfoSection(order)

            // Billing Information
            billingInfoSection(order)

            // Timestamps
            timestampsSection(order)
          }
          .padding()
        }
      }
      .navigationTitle("Order Details")
      .navigationBarTitleDisplayMode(.large)
      .toolbar {
        ToolbarItem(placement: .navigationBarTrailing) {
          HStack(spacing: 16) {
            Button("Edit") { onEditTapped() }
              .disabled(isLoading)
            Button("Done") { dismiss() }
          }
        }
      }
    }
    .onAppear {
      loadOrderDetails()
    }
  }

  private func onEditTapped() {
    isLoading = true
    FirebaseOrderManager.shared.fetchOrderDocument(orderId: orderId) { result in
      DispatchQueue.main.async {
        self.isLoading = false
        switch result {
        case .failure(let error):
          self.errorMessage = error.localizedDescription
        case .success(let data):
          let status =
            (data["status"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased() ?? ""
          let step = data["currentStep"] as? Int ?? 1

          if status == "completed" {
            AppStoreLinking.openYouTube()
            return
          }

          // Resume editing this order at its saved step
          navigationState.resumeOrder(orderId: orderId, at: max(1, min(6, step)))
          // Mark that we've applied resume for this order so it won't be reapplied unexpectedly
          navigationState.lastAppliedResumeForOrderId = orderId
          // Dismiss on the next runloop cycle to avoid SwiftUI double-pop navigation behavior
          DispatchQueue.main.async {
            dismiss()
          }
        }
      }
    }
  }

  // Centralized deep link logic lives in AppStoreLinking

  private func orderHeaderSection(_ order: OrderDetail) -> some View {
    VStack(spacing: 8) {
      HStack {
        Text("Order ID")
          .font(.caption)
          .foregroundColor(.secondary)
        Spacer()
        Text(order.orderId)
          .font(.caption)
          .fontWeight(.medium)
      }

      HStack {
        Text("Status")
          .font(.headline)
        Spacer()
        Text(order.status.capitalized)
          .font(.headline)
          .foregroundColor(statusColor(order.status))
          .padding(.horizontal, 12)
          .padding(.vertical, 4)
          .background(statusColor(order.status).opacity(0.1))
          .cornerRadius(8)
      }
    }
    .padding()
    .background(Color.adaptiveSecondaryBackground)
    .cornerRadius(10)
  }

  private func personalInfoSection(_ order: OrderDetail) -> some View {
    sectionCard(title: "Personal Information") {
      detailRow(label: "Name", value: "\(order.firstName) \(order.lastName)")
      detailRow(label: "Email", value: order.email)
      if !order.phoneNumber.isEmpty {
        detailRow(label: "Phone", value: order.phoneNumber)
      }
    }
  }

  private func addressInfoSection(_ order: OrderDetail) -> some View {
    sectionCard(title: "Address Information") {
      detailRow(label: "Street", value: order.street)
      if !order.aptNumber.isEmpty {
        detailRow(label: "Apt/Unit", value: order.aptNumber)
      }
      detailRow(label: "City", value: order.city)
      detailRow(label: "State", value: order.state)
      detailRow(label: "ZIP", value: order.zip)
      detailRow(label: "Country", value: order.country)
    }
  }

  private func deviceInfoSection(_ order: OrderDetail) -> some View {
    sectionCard(title: "Device Information") {
      detailRow(label: "Brand", value: order.deviceBrand)
      detailRow(label: "Model", value: order.deviceModel)
      detailRow(label: "Compatible", value: order.deviceIsCompatible ? "Yes" : "No")
      detailRow(label: "For This Device", value: order.isForThisDevice ? "Yes" : "No")
      if !order.imei.isEmpty {
        detailRow(label: "IMEI", value: order.imei)
      }
    }
  }

  private func serviceInfoSection(_ order: OrderDetail) -> some View {
    sectionCard(title: "Service Information") {
      detailRow(label: "Number Type", value: order.numberType)
      if !order.selectedPhoneNumber.isEmpty {
        detailRow(label: "Selected Number", value: order.selectedPhoneNumber)
      }
      detailRow(label: "SIM Type", value: order.simType)
      detailRow(label: "Port-in Skipped", value: order.portInSkipped ? "Yes" : "No")
      detailRow(label: "Show QR Code", value: order.showQRCode ? "Yes" : "No")
    }
  }

  private func billingInfoSection(_ order: OrderDetail) -> some View {
    sectionCard(title: "Billing Information") {
      if !order.billingDetails.isEmpty {
        detailRow(label: "Billing Details", value: order.billingDetails)
      }
      if !order.creditCardNumber.isEmpty {
        detailRow(label: "Credit Card", value: maskCreditCard(order.creditCardNumber))
      }
    }
  }

  private func timestampsSection(_ order: OrderDetail) -> some View {
    sectionCard(title: "Order Timeline") {
      detailRow(label: "Created", value: dateFormatter.string(from: order.createdAt))
      detailRow(label: "Last Updated", value: dateFormatter.string(from: order.updatedAt))
      if let completionDate = order.orderCompletionDate {
        detailRow(
          label: "Completed", value: dateFormatter.string(from: completionDate))
      }
    }
  }

  private func sectionCard<Content: View>(
    title: String, @ViewBuilder content: () -> Content
  ) -> some View {
    VStack(alignment: .leading, spacing: 12) {
      Text(title)
        .font(.headline)
        .fontWeight(.semibold)

      VStack(spacing: 8) {
        content()
      }
    }
    .padding()
    .background(Color.adaptiveBackground)
    .cornerRadius(10)
    .shadow(color: .black.opacity(0.05), radius: 2, x: 0, y: 1)
  }

  private func detailRow(label: String, value: String) -> some View {
    HStack {
      Text(label)
        .font(.subheadline)
        .foregroundColor(.secondary)
      Spacer()
      Text(value)
        .font(.subheadline)
        .fontWeight(.medium)
        .multilineTextAlignment(.trailing)
    }
  }

  private func statusColor(_ status: String) -> Color {
    switch status.lowercased() {
    case "completed":
      return .green
    case "pending":
      return .orange
    case "cancelled":
      return .red
    case "processing":
      return .blue
    default:
      return .secondary
    }
  }

  private func maskCreditCard(_ cardNumber: String) -> String {
    if cardNumber.count <= 4 {
      return cardNumber
    }
    let lastFour = String(cardNumber.suffix(4))
    let masked = String(repeating: "*", count: max(0, cardNumber.count - 4))
    return masked + lastFour
  }

  private func loadOrderDetails() {
    guard let userId = Auth.auth().currentUser?.uid else {
      DispatchQueue.main.async {
        self.isLoading = false
        self.errorMessage = "User not authenticated"
      }
      return
    }

    let db = Firestore.firestore()

    db.collection("users").document(userId)
      .collection("orders").document(orderId)
      .getDocument { document, error in
        DispatchQueue.main.async {
          self.isLoading = false

          if let error = error {
            self.errorMessage = "Failed to load order details: \(error.localizedDescription)"
            return
          }

          guard let document = document, document.exists,
            let data = document.data()
          else {
            self.errorMessage = "Order not found"
            return
          }

          self.orderDetail = OrderDetail(from: data)
        }
      }
  }
}

#Preview {
  OrderDetailView(orderId: "sample-order-id")
}

File: linkmobile/lib/screens/order_flow/order_detail_view.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:cloud_firestore/cloud_firestore.dart';
import '../../models/order_models.dart';
import '../../providers/navigation_state.dart';
import '../../services/firebase_order_manager.dart';
import '../../utils/theme.dart';

class OrderDetailView extends StatefulWidget {
  final String orderId;

  const OrderDetailView({
    super.key,
    required this.orderId,
  });

  @override
  State<OrderDetailView> createState() => _OrderDetailViewState();
}

class _OrderDetailViewState extends State<OrderDetailView> {
  OrderDetail? _orderDetail;
  Map<String, dynamic>? _orderData;
  bool _isLoading = true;
  String? _errorMessage;

  @override
  void initState() {
    super.initState();
    _loadOrderDetails();
  }

  Future<void> _loadOrderDetails() async {
    final user = FirebaseAuth.instance.currentUser;
    if (user == null) {
      setState(() {
        _isLoading = false;
        _errorMessage = 'User not authenticated';
      });
      return;
    }

    try {
      final orderManager = FirebaseOrderManager();
      final orderData = await orderManager.fetchOrderDocument(user.uid, widget.orderId);
      
      if (orderData == null) {
        setState(() {
          _isLoading = false;
          _errorMessage = 'Order not found';
        });
        return;
      }

      setState(() {
        _orderData = orderData;
        _orderDetail = OrderDetail.fromMap(orderData);
        _isLoading = false;
      });
    } catch (e) {
      setState(() {
        _isLoading = false;
        _errorMessage = 'Failed to load order details: ${e.toString()}';
      });
    }
  }

  void _onEditTapped() async {
    setState(() {
      _isLoading = true;
    });

    try {
      final user = FirebaseAuth.instance.currentUser;
      if (user == null) {
        setState(() {
          _isLoading = false;
          _errorMessage = 'User not authenticated';
        });
        return;
      }

      final orderManager = FirebaseOrderManager();
      final orderData = await orderManager.fetchOrderDocument(user.uid, widget.orderId);
      
      if (orderData == null) {
        setState(() {
          _isLoading = false;
          _errorMessage = 'Order not found';
        });
        return;
      }

      final status = (orderData['status'] as String? ?? '').trim().toLowerCase();
      final step = orderData['currentStep'] as int? ?? 1;

      if (status == 'completed') {
        // Handle completed order - could open YouTube or show message
        if (mounted) {
          ScaffoldMessenger.of(context).showSnackBar(
            const SnackBar(content: Text('This order is already completed')),
          );
        }
        setState(() {
          _isLoading = false;
        });
        return;
      }

      // Resume editing this order at its saved step
      final navigationState = Provider.of<NavigationState>(context, listen: false);
      navigationState.resumeOrder(widget.orderId, step.clamp(1, 6));
      navigationState.lastAppliedResumeForOrderId = widget.orderId;
      
      // Navigate back - ContentView will detect the navigation state change
      // and show the order flow at the correct step
      if (mounted) {
        Navigator.of(context).pop();
      }
    } catch (e) {
      setState(() {
        _isLoading = false;
        _errorMessage = 'Failed to load order: ${e.toString()}';
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Order Details'),
        backgroundColor: Colors.transparent,
        elevation: 0,
        flexibleSpace: Container(
          decoration: BoxDecoration(
            gradient: AppTheme.blueGradient,
          ),
        ),
        actions: [
          TextButton(
            onPressed: _isLoading ? null : _onEditTapped,
            style: TextButton.styleFrom(
              foregroundColor: Colors.white,
            ),
            child: const Text('Edit'),
          ),
          TextButton(
            onPressed: () => Navigator.of(context).pop(),
            style: TextButton.styleFrom(
              foregroundColor: Colors.white,
            ),
            child: const Text('Done'),
          ),
        ],
      ),
      body: _isLoading
          ? const Center(child: CircularProgressIndicator())
          : _errorMessage != null
              ? Center(
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: [
                      const Icon(
                        Icons.error_outline,
                        size: 64,
                        color: Colors.red,
                      ),
                      const SizedBox(height: 16),
                      Text(
                        'Error',
                        style: Theme.of(context).textTheme.titleLarge,
                      ),
                      const SizedBox(height: 8),
                      Padding(
                        padding: const EdgeInsets.all(16.0),
                        child: Text(
                          _errorMessage!,
                          textAlign: TextAlign.center,
                        ),
                      ),
                    ],
                  ),
                )
              : _orderDetail != null && _orderData != null
                  ? SingleChildScrollView(
                      padding: const EdgeInsets.all(16.0),
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.stretch,
                        children: [
                          _buildOrderHeaderSection(),
                          const SizedBox(height: 16),
                          _buildPersonalInfoSection(_orderDetail!),
                          const SizedBox(height: 16),
                          _buildAddressInfoSection(_orderDetail!),
                          const SizedBox(height: 16),
                          _buildDeviceInfoSection(_orderDetail!),
                          const SizedBox(height: 16),
                          _buildServiceInfoSection(_orderDetail!),
                          const SizedBox(height: 16),
                          _buildBillingInfoSection(_orderDetail!),
                        ],
                      ),
                    )
                  : const Center(child: Text('No order data')),
    );
  }

  Widget _buildOrderHeaderSection() {
    final status = _orderData!['status'] as String? ?? 'pending';
    return Container(
      padding: const EdgeInsets.all(16.0),
      decoration: BoxDecoration(
        color: Colors.grey[100],
        borderRadius: BorderRadius.circular(10),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: [
              Text(
                'Order ID',
                style: Theme.of(context).textTheme.bodySmall,
              ),
              Text(
                widget.orderId.length > 8 ? widget.orderId.substring(0, 8) : widget.orderId,
                style: Theme.of(context).textTheme.bodySmall?.copyWith(
                      fontWeight: FontWeight.w600,
                    ),
              ),
            ],
          ),
          const SizedBox(height: 8),
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: [
              Text(
                'Status',
                style: Theme.of(context).textTheme.titleMedium?.copyWith(
                      fontWeight: FontWeight.w600,
                    ),
              ),
              Container(
                padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 4),
                decoration: BoxDecoration(
                  color: _getStatusColor(status).withOpacity(0.1),
                  borderRadius: BorderRadius.circular(8),
                ),
                child: Text(
                  status.toUpperCase(),
                  style: TextStyle(
                    color: _getStatusColor(status),
                    fontWeight: FontWeight.w600,
                  ),
                ),
              ),
            ],
          ),
        ],
      ),
    );
  }

  Widget _buildPersonalInfoSection(OrderDetail order) {
    return _buildSectionCard(
      title: 'Personal Information',
      children: [
        _buildDetailRow('Name', '${order.firstName ?? ''} ${order.lastName ?? ''}'.trim()),
        _buildDetailRow('Email', order.email ?? 'N/A'),
        if (order.phoneNumber != null && order.phoneNumber!.isNotEmpty)
          _buildDetailRow('Phone', order.phoneNumber!),
      ],
    );
  }

  Widget _buildAddressInfoSection(OrderDetail order) {
    return _buildSectionCard(
      title: 'Address Information',
      children: [
        _buildDetailRow('Street', order.street ?? 'N/A'),
        if (order.aptNumber != null && order.aptNumber!.isNotEmpty)
          _buildDetailRow('Apt/Unit', order.aptNumber!),
        _buildDetailRow('City', order.city ?? 'N/A'),
        _buildDetailRow('State', order.state ?? 'N/A'),
        _buildDetailRow('ZIP', order.zip ?? 'N/A'),
        _buildDetailRow('Country', order.country ?? 'N/A'),
      ],
    );
  }

  Widget _buildDeviceInfoSection(OrderDetail order) {
    return _buildSectionCard(
      title: 'Device Information',
      children: [
        _buildDetailRow('Brand', order.deviceBrand ?? 'N/A'),
        _buildDetailRow('Model', order.deviceModel ?? 'N/A'),
        _buildDetailRow(
          'Compatible',
          order.deviceIsCompatible == true ? 'Yes' : 'No',
        ),
        if (order.imei != null && order.imei!.isNotEmpty)
          _buildDetailRow('IMEI', order.imei!),
      ],
    );
  }

  Widget _buildServiceInfoSection(OrderDetail order) {
    return _buildSectionCard(
      title: 'Service Information',
      children: [
        _buildDetailRow('Number Type', order.numberType ?? 'N/A'),
        if (order.selectedPhoneNumber != null &&
            order.selectedPhoneNumber!.isNotEmpty)
          _buildDetailRow('Selected Number', order.selectedPhoneNumber!),
        _buildDetailRow('SIM Type', order.simType ?? 'N/A'),
        _buildDetailRow(
          'Port-in Skipped',
          order.portInSkipped == true ? 'Yes' : 'No',
        ),
      ],
    );
  }

  Widget _buildBillingInfoSection(OrderDetail order) {
    return _buildSectionCard(
      title: 'Billing Information',
      children: [
        if (order.billingDetails != null && order.billingDetails!.isNotEmpty)
          _buildDetailRow('Billing Details', order.billingDetails!),
        if (order.creditCardNumber != null &&
            order.creditCardNumber!.isNotEmpty)
          _buildDetailRow(
            'Credit Card',
            _maskCreditCard(order.creditCardNumber!),
          ),
      ],
    );
  }

  Widget _buildSectionCard({
    required String title,
    required List<Widget> children,
  }) {
    return Container(
      padding: const EdgeInsets.all(16.0),
      decoration: BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.circular(10),
        boxShadow: [
          BoxShadow(
            color: Colors.black.withOpacity(0.05),
            blurRadius: 2,
            offset: const Offset(0, 1),
          ),
        ],
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          Text(
            title,
            style: Theme.of(context).textTheme.titleMedium?.copyWith(
                  fontWeight: FontWeight.w600,
                ),
          ),
          const SizedBox(height: 12),
          ...children,
        ],
      ),
    );
  }

  Widget _buildDetailRow(String label, String value) {
    return Padding(
      padding: const EdgeInsets.only(bottom: 8.0),
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Expanded(
            flex: 2,
            child: Text(
              label,
              style: Theme.of(context).textTheme.bodySmall?.copyWith(
                    color: Colors.grey[600],
                  ),
            ),
          ),
          Expanded(
            flex: 3,
            child: Text(
              value,
              textAlign: TextAlign.end,
              style: Theme.of(context).textTheme.bodySmall?.copyWith(
                    fontWeight: FontWeight.w500,
                  ),
            ),
          ),
        ],
      ),
    );
  }

  Color _getStatusColor(String status) {
    switch (status.toLowerCase()) {
      case 'completed':
        return Colors.green;
      case 'pending':
        return Colors.orange;
      case 'cancelled':
        return Colors.red;
      case 'draft':
        return Colors.grey;
      case 'processing':
        return AppTheme.mainBlue;
      default:
        return Colors.grey;
    }
  }

  String _maskCreditCard(String cardNumber) {
    if (cardNumber.length <= 4) {
      return cardNumber;
    }
    final lastFour = cardNumber.substring(cardNumber.length - 4);
    final masked = '*' * (cardNumber.length - 4);
    return '$masked$lastFour';
  }
}

